stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +221,188 @@ class task_system {
         x = std::move(_q.front());
         _q.pop_front();
         return true;
     }
 
+    bool pop(task<void()>& x) {
+        lock_t lock{_queue_mutex};
+        if (_q.empty()) return false;
+        x = std::move(_q.front());
+        _q.pop_front();
+        return true;
+    }
+
+
     template <typename F>
     bool try_push(F&& f) {
         {
-            lock_t lock{_mutex, std::try_to_lock};
+            lock_t lock{_queue_mutex, std::try_to_lock};
             if (!lock) return false;
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
-        return true;
-    }
-
-    void done() {
-        {
-            lock_t lock{_mutex};
-            _done = true;
-        }
-        _ready.notify_all();
-    }
-
-    bool pop(task<void()>& x) {
-        lock_t lock{_mutex};
-        while (_q.empty() && !_done)
-            _ready.wait(lock);
-        if (_q.empty()) return false;
-        x = std::move(_q.front());
-        _q.pop_front();
+        _ready->notify_all();
         return true;
     }
 
     template <typename F>
     void push(F&& f) {
         {
-            lock_t lock{_mutex};
+            lock_t lock{_queue_mutex};
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
+        _ready->notify_all();
     }
 };
 
 /**************************************************************************************************/
 
-class task_system {
+class priority_task_system {
+    using lock_t = std::unique_lock<std::mutex>;
+
     const unsigned _count{std::thread::hardware_concurrency()};
-    std::vector<std::thread> _threads;
-    std::vector<notification_queue> _q{_count};
+
+    struct thread_context
+    {
+        std::mutex mutex;
+        std::condition_variable ready;
+        std::thread thread;
+    };
+    std::vector<thread_context> _thread_contexts{_count};
+    std::array<std::vector<notification_queue>, 3> _q;
     std::atomic<unsigned> _index{0};
+    std::atomic_bool _done{false};
 
     void run(unsigned i) {
         while (true) {
+            begin:
             task<void()> f;
 
-            for (unsigned n = 0; n != _count * 32; ++n) {
-                if (_q[(i + n) % _count].try_pop(f)) break;
+            for (auto& q : _q) {
+                for (unsigned n = 0; n != 64 / _count; ++n) {
+                    if (q[(i + n) % _count].try_pop(f)) {
+                        f();
+                        goto begin;
+                    };
+                }
+            }
+
+            for (auto& q : _q) {
+                if (q[i].pop(f)) {
+                    f();
+                    goto begin;
+                }
             }
-            if (!f && !_q[i].pop(f)) break;
 
-            f();
+            {
+                lock_t lock{_thread_contexts[i].mutex};
+                while(!_done) {
+                    goto begin;
+                    _thread_contexts[i].ready.wait(lock);
```

Comment:
@FelixPetriconi Did you mean to leave `goto begin;` here? It looks like this code is unreachable: `_thread_contexts[i].ready.wait(lock);`

Reply:
I had to read the code a bit to figure this one out, too. The wait could either be spurious or come from `notification_queue`. On that note: when a spurious unblock happens, it's still OK to return to the top of this routine?

Reply:
I fixed the order in a later commit. Since the spurious wake ups happen very seldom from my experience, it is fine to jump to the beginning and in worst case run into the next wait.

Reply:
In this case, then, could the `while` be an `if`?